### PR TITLE
Implement label_replace().

### PR DESCRIPTION
**Pull request by juliusv**
Implements part of https://github.com/prometheus/prometheus/issues/959.

### CODE REVIEW

**Comment by juliusv** [comment]
@brian-brazil @fabxc 

This currently does sub-string matches, like we do in label selectors and relabeling rules. We might want to switch all of them over to full string matches at some point. See also https://github.com/prometheus/prometheus/issues/996.

File: promql/functions.go [hunk 1]
```diff
@@ -603,6 +604,40 @@ func funcChanges(ev *evaluator, args Expressions) Value {
...
+	}
+
+	for _, el := range vector {
+		srcVal := string(el.Metric.Metric[src])
+		indexes := regex.FindStringSubmatchIndex(srcVal)
+		// If there is no match, no replacement should take place.
+		if indexes == nil {
+			continue
+		}
+		regex.ExpandString([]byte{}, repl, srcVal, indexes)
```
Comment (brian-brazil) [inline promql/functions.go]: Dupe?

File: promql/functions.go [hunk 2]
```diff
@@ -603,6 +604,40 @@ func funcChanges(ev *evaluator, args Expressions) Value {
 	return out
 }
 
+// === label_replace(vector ExprVector, dst_label, replacement, src_labelname, regex ExprString) Vector ===
+func funcLabelReplace(ev *evaluator, args Expressions) Value {
+	var (
+		vector   = ev.evalVector(args[0])
+		dst      = clientmodel.LabelName(ev.evalString(args[1]).Value)
```
Comment (brian-brazil) [inline promql/functions.go]: When could this not be a string?

File: promql/testdata/functions.test [hunk 3]
```diff
@@ -71,24 +71,63 @@ load 5m
...
+eval instant at 0m label_replace(testmetric, "dst", "destination-value-$1", "src", "value-(.*)")
+  testmetric{src="source-value-10",dst="destination-value-10"} 0
+  testmetric{src="source-value-20",dst="destination-value-20"} 1
+
+# label_replace works with multiple capture groups.
+eval instant at 0m label_replace(testmetric, "dst", "$1-value-$2", "src", "(.*)-value-(.*)")
+  testmetric{src="source-value-10",dst="source-value-10"} 0
+  testmetric{src="source-value-20",dst="source-value-20"} 1
+
+# label_replace does not overwrite the destination label if the source label
```
Comment (brian-brazil) [inline promql/testdata/functions.test]: It should work if the source label is empty, and the regex matches it. Can you add a test for this?

File: promql/functions.go (hunk 1 above)
Comment (juliusv) [inline promql/functions.go]: Oops, removed.

File: promql/functions.go (hunk 2 above)
Comment (juliusv) [inline promql/functions.go]: It's always a string, but we need to eval it from being a string expression node into its actual string value first via `evalString`.

File: promql/testdata/functions.test (hunk 3 above)
Comment (juliusv) [inline promql/testdata/functions.test]: Added a test for that.

**Comment by juliusv** [comment]
Adding the last test made me think: what if relabeling removes a distinguishing label between time series, so that the output vector will contain the same series multiple times? Should we keep track of that by keeping a hash table of all output label sets and erroring if a labelset appears more than once?

**Comment by fabxc** [comment]
Yes, that's an error.

File: promql/testdata/functions.test [hunk 4]
```diff
@@ -71,24 +71,69 @@ load 5m
...
+  testmetric{src="source-value-20",dst="original-destination-value"} 1
+
+# label_replace drops labels that are set to empty values.
+eval instant at 0m label_replace(testmetric, "dst", "", "dst", "")
+  testmetric{src="source-value-10"} 0
+  testmetric{src="source-value-20"} 1
+
+# label_replace fails when the regex is invalid.
+eval_fail instant at 0m label_replace(testmetric, "dst", "value-$1", "src", "(.*")
+  testmetric{src="source-value-10",dst="original-destination-value"} 0
```
Comment (brian-brazil) [inline promql/testdata/functions.test]: Why do you need the result sample for failure?

**Comment by brian-brazil** [comment]
Agreed, we should treat that as an error like we do for binops.

File: promql/testdata/functions.test (hunk 4 above)
Comment (juliusv) [inline promql/testdata/functions.test]: Oh yeah, that's removed in the latest push.

**Comment by brian-brazil** [comment]
:+1:

**Comment by juliusv** [comment]
Ok, I added duplicate detection, and also a check that the destination label name is actually valid. I omitted a check for validity of the source label name, as an invalid source shouldn't matter, but don't feel strongly whether it should be added or not.

**Pull request merged**